File: 프로그래머스/unrated/172928. 공원 산책/공원 산책.java
class Solution {
    public int[] solution(String[] park, String[] routes) {
        int[] answer = {};
        for(int i = 0; i < park.length; i++){//시작점 찾기.
            if(park[i].contains("S")){
                int j = park[i].indexOf("S");
                answer = new int[] {i,j};
            }
        }
        
loop:   for(int i = 0 ; i < routes.length; i++){
            int record = routes[i].charAt(2)-'0';//1부터 9
            switch(routes[i].charAt(0)){
                case 'N':  //위
                    int tmp = record;
                    while(tmp > 0){
                        if(answer[0]-tmp < 0 || park[answer[0]-tmp].charAt(answer[1])=='X') continue loop;
                        tmp--;//위의 과정들을 다 뚫었다면 값 갱신
                    }
                    answer[0] -= record;
                    break;
                case 'S':  //아
                    tmp = record;
                while(tmp > 0){
                    if(answer[0] + tmp >= park.length || park[answer[0] + tmp].charAt(answer[1])=='X') continue loop;
                    tmp--;//위의 과정들을 다 뚫었다면 값 갱신
                }
                answer[0] += record;
                    break;
                case 'W':  //왼
                    tmp = record;
                while(tmp > 0){
                    if(answer[1]-tmp < 0 || park[answer[0]].charAt(answer[1]-tmp)=='X') continue loop;
                    tmp--;//위의 과정들을 다 뚫었다면 값 갱신
                }
                answer[1] -= record;
                    break;
                case 'E':  //오
                    tmp = record;
                while(tmp > 0){
                    if(answer[1] + tmp >= park[answer[0]].length() || park[answer[0]].charAt(answer[1] + tmp)=='X') continue loop;
                    tmp--;//위의 과정들을 다 뚫었다면 값 갱신
                }
                answer[1] += record;
                    break;
            }           
        }

        return answer;
    }
}

File: 프로그래머스/1/172928. 공원 산책/공원 산책.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

const int dr[] = {-1, 0, 1, 0};
const int dc[] = {0, 1, 0, -1};
int dd[128];
int R, C, r, c;
char** P;

void move(int d, int s){
    int nr = r, nc = c;
    while(s--){
        nr += dr[d], nc += dc[d];
        if(nr < 0 || nr >= R || nc < 0 || nc >= C) return;
        if(P[nr][nc] == 'X') return;
    }
    r = nr, c = nc;
}

int* solution(const char* park[], size_t park_len, const char* routes[], size_t routes_len) {
    dd['N'] = 0, dd['E'] = 1, dd['S'] = 2, dd['W'] = 3;
    int* answer = (int*)malloc(2 * sizeof(int));
    R = park_len, C = strlen(park[0]); P = park;
    int flag = 0;
    for(r = 0; r < R; r++){
        for(c = 0; c < C; c++){
            if(P[r][c] == 'S'){
                flag = 1;
                break;
            }
        }
        if(flag) break;
    }
    for(int i = 0; i < routes_len; i++){
        move(dd[routes[i][0]], routes[i][2] - '0');
    }
    answer[0] = r, answer[1] = c;
    return answer;
}